traversal/expression/Operation.java

```diff
@@ -0,0 +1,101 @@
+            case POW:   return instantiatePow(first, second);
+
+            default:
+                throw TypeDBException.of(ILLEGAL_STATE);
+        }
+    }
+
+    private static Expression<?> instantiatePlus(Expression<?> first, Expression<?> second) {
+        if (first.isLong()) {
+            if (second.isLong()) return new ExpressionBase.Binary.Long<>(first.asLong(), second.asLong(), (a,b) -> a + b);
+            if (second.isDouble()) return new ExpressionBase.Binary.Double<>(first.asLong(), second.asDouble(), (a,b) -> a.doubleValue() + b);
+        }
+
+        if (first.isDouble()) {
+            if (second.isLong()) return new ExpressionBase.Binary.Double<>(first.asDouble(), second.asLong(), (a,b) -> a + b.doubleValue());
+            if (second.isDouble()) return new ExpressionBase.Binary.Double<>(first.asDouble(), second.asDouble(), (a,b) -> a + b);
+        }
+
+        throw TypeDBException.of(OPERATION_NOT_FOUND,
+                first.returnType().name(), TypeQLToken.Expression.Operation.PLUS, second.returnType().name());
+    }
+
+    private static Expression<?> instantiateMinus(Expression<?> first, Expression<?> second) {
+
+        if (first.isLong()) {
+            if (second.isLong()) return new ExpressionBase.Binary.Long<>(first.asLong(), second.asLong(), (a,b) -> a - b);
+            if (second.isDouble()) return new ExpressionBase.Binary.Double<>(first.asLong(), second.asDouble(), (a,b) -> a.doubleValue() - b);
+        }
+
+        if (first.isDouble()) {
+            if (second.isLong()) return new ExpressionBase.Binary.Double<>(first.asDouble(), second.asLong(), (a,b) -> a - b.doubleValue());
+            if (second.isDouble()) return new ExpressionBase.Binary.Double<>(first.asDouble(), second.asDouble(), (a,b) -> a - b);
+        }
+
+        throw TypeDBException.of(OPERATION_NOT_FOUND,
+                first.returnType().name(), TypeQLToken.Expression.Operation.MINUS, second.returnType().name());
+    }
+    private static Expression<?> instantiateTimes(Expression<?> first, Expression<?> second) {
+
+        if (first.isLong()) {
+            if (second.isLong()) return new ExpressionBase.Binary.Long<>(first.asLong(), second.asLong(), (a,b) -> a * b);
+            if (second.isDouble()) return new ExpressionBase.Binary.Double<>(first.asLong(), second.asDouble(), (a,b) -> a.doubleValue() * b);
+        }
+
+        if (first.isDouble()) {
+            if (second.isLong()) return new ExpressionBase.Binary.Double<>(first.asDouble(), second.asLong(), (a,b) -> a * b.doubleValue());
+            if (second.isDouble()) return new ExpressionBase.Binary.Double<>(first.asDouble(), second.asDouble(), (a,b) -> a * b);
+        }
+
+        throw TypeDBException.of(OPERATION_NOT_FOUND,
+                first.returnType().name(), TypeQLToken.Expression.Operation.TIMES, second.returnType().name());
+    }
+
+    private static Expression<?> instantiateDiv(Expression<?> first, Expression<?> second) {
```

Comment:
Do we want to throw our own exception rather than Java's for a Division by 0 error? Then we can show which expression caused it, etc.
Secondly, we should add a test for division by 0 to BDD

Reply:
We should. But I'd rather do it higher up so there's less chance of missing them.

Reply:
Turns out we do have a BDD test for division by 0 and we do throw our own error.
But it's specifically handled by the check in `traversal.expression.Expression.Double.evaluateToVertex`

I'd like to handle all errors around there